### ISSUE TITLE
Query syntax for deep populate/select?

### ISSUE THREAD

**Issue by timbur** [wprl/baucis#254]
Suppose some schema looks like this:
```js
var CommentSchema = new Schema({
  created: {
    type: Date,
    default: Date.now,
    required: true
  },
  createdBy: {
    type: ObjectId,
    ref: 'User',
    required: true
  },
  modified: {
    type: Date,
    default: Date.now,
    required: true
  },
  modifiedBy: {
    type: ObjectId,
    ref: 'User',
    required: true
  },
  comment: {
    type: String,
    required: true
  },
  replies: {
    type: [{
      type: ObjectId,
      ref: 'Comment'
    }]
  }
});
```

And suppose we want to go 3 levels deep and select only certain fields.  Is something like this supported?
```js
var select = 'created createdBy comment replies';
var populate = function (level) {
  var replies = {
    path: "replies",
    select: select
  };

  if (level > 0) {
    replies.populate = populate(level - 1);
  }

  return [
    "createdBy",
    replies
  ];
};

var query = {
  conditions: {
    createdBy: "user123"
  },
  select: select,
  populate: populate(3)
};

fetch('/api/comments?'+JSON.stringify(query)).then(console.log.bind(console));
// /api/comments?{"conditions":{"createdBy":"user123"},"select":"created createdBy comment replies","populate":["createdBy",{"path":"replies","select":"created createdBy comment replies","populate":["createdBy",{"path":"replies","select":"created createdBy comment replies","populate":["createdBy",{"path":"replies","select":"created createdBy comment replies","populate":["createdBy",{"path":"replies","select":"created createdBy comment replies"}]}]}]}]}
```

This particular example could obviously be simplified with some parameter designed for it since it's all just the same nested query, but my real question is: does something like this work with many different nested types of references?  If not, any recommendations on how to implement it?

**Comment by timbur** [comment]
Actually, that was a bad example since it didn't describe deep population.  Let me update it real quick.

**Comment by timbur** [comment]
Updated OP.

**Comment by timbur** [comment]
I should probably just match the syntax of `mongoose-deep-populate` as closely as possible within the query URL, but I'm still a bit unsure about the `select` syntax if/when I want to select different fields at different levels.

Something like this seems logical to me (note the lack of the `populate` parameter, as that should be implied via the `select` statement):
```
api/comments?select=created createdBy.followers replies.createdBy.followers replies.createdBy.name replies.replies.createdBy.name replies.replies.replies.createdBy.name
```

Hmm... I may have just answered my own question and it might work with minimal effort, but I am curious to know if anyone else has experience doing something like this.

<sup>Sorry for all the posts talking to myself.  xD</sup>

**Comment by timbur** [comment]
Alright so... I first tried moving everything into mongoose plugins instead of decorating the baucis controllers, but it ended up being too hacky to pass the Express request object into the mongoose plugins, so I then moved everything back into the baucis controllers.  It would seem the best solution is to keep track of which fields are being populated/selected and then go deep into the outgoing docs to apply the custom decorator(s) logic.  I'll create a repo with all my code at some point if anyone is interested.
